astropy/table/index.py

```diff
@@ -0,0 +1,630 @@
+from copy import deepcopy
+import numpy as np
+
+from .bst import BST, FastBST, FastRBT, MinValue, MaxValue
+from .sorted_array import SortedArray
+
+'''
+The Index class can use several implementations as its
+engine. Any implementation should implement the following:
+
+__init__([dict] lines) : initializes based on key/row list pairs
+add(key, row) -> None : add (key, row) to existing data
+remove(key, data=None) -> boolean : remove data from self[key], or all of
+                                    self[key] if data is None
+shift_left(row) -> None : decrement row numbers after row
+shift_right(row) -> None : increase row numbers >= row
+find(key) -> list : list of rows corresponding to key
+range(lower, upper, bounds) -> list : rows in self[k] where k is between
+                               lower and upper (<= or < based on bounds)
+sort() -> list of rows in sorted order (by key)
+replace_rows(row_map) -> None : replace row numbers based on slice
+items() -> list of tuples of the form (key, data)
+
+Notes
+-----
+    When a Table is initialized from another Table, indices are
+    (deep) copied and their columns are set to the columns of the new Table.
+
+    Column creation:
+    Column(c) -> deep copy of indices
+    c[[1, 2]] -> deep copy and reordering of indices
+    c[1:2] -> reference
+    array.view(Column) -> no indices
+'''
+
+class QueryError(ValueError):
+    '''
+    Indicates that a given index cannot handle the supplied query.
+    '''
+    pass
+
+class Index(object):
+    '''
+    The Index class makes it possible to maintain indices
+    on columns of a Table, so that column values can be queried
+    quickly and efficiently. Column values are stored in lexicographic
+    sorted order, which allows for binary searching in O(log n).
+
+    Parameters
+    ----------
+    columns : list or None
+        List of columns on which to create an index. If None,
+        create an empty index for purposes of deep copying.
+    impl : type or None
+        Indexing engine class to use, from among SortedArray, BST,
+        FastBST, and FastRBT. If the supplied argument is None (by
+        default), use SortedArray.
+    col_dtypes : list or None
+        List of dtypes to use for each column
+    data : SortedArray, BST, FastBST, FastRBT, or None
+        Engine data to copy
+    '''
+    def __new__(cls, *args, **kwargs):
+        self = super(Index, cls).__new__(cls)
+        self.__init__(*args, **kwargs)
+        return SlicedIndex(self, slice(0, 0, None), original=True)
```

Comment:
Ah, that's because for "original" slices, the `stop` property is determined by actual column length rather than the input slice (which determines `_stop`). Now that I think about it, it might make more sense to keep `stop` as an ordinary variable and modify it whenever column length changes.